iml-agent/src/action_plugins/stratagem/action_filesync.rs

```diff
@@ -0,0 +1,208 @@
+                let mut rsync_res = do_rsync(&rsync_list).await?;
+
+                result.append(&mut rsync_res);
+                rsync_res.clear();
+                rsync_list.clear();
+            }
+        }
+    }
+
+    let mut rsync_res = do_rsync(&rsync_list).await?;
+
+    result.append(&mut rsync_res);
+
+    Ok(result)
+}
+
+async fn restore_fids(
+    llapi: LlapiFid,
+    dest_src: &String,
+    fid_list: Vec<FidItem>,
+) -> Result<Vec<FidError>, ImlAgentError> {
+    let mut result = vec![];
+
+    for fid in fid_list {
+        let fid_path = llapi.fid2path(&fid.fid)?;
+        let src_file = format!("{}/{}", llapi.mntpt(), &fid_path);
+        let dest_file = format!("{}/{}", &dest_src, &fid_path);
+        let mut dest_dir = PathBuf::from(&src_file);
+
+        dest_dir.pop();
+        fs::create_dir_all(&dest_dir).await?;
+
+        let md = fs::metadata(&dest_file).await?;
+        /* invoking mpifileutils is slow, so only do it for directories
+         * and big files, otherwise just use rsync
+         */
+        let output;
+        if md.is_dir() || (md.len() > 1024 * 1024 * 1024) {
+            output = Command::new("mpirun")
+                .arg("--hostfile")
+                .arg("/etc/iml/filesync-hostfile")
+                .arg("--allow-run-as-root")
+                .arg("dsync")
+                .arg("-S")
+                .arg(dest_file)
+                .arg(src_file)
+                .output();
+        } else {
+            output = Command::new("rsync").arg(dest_file).arg(src_file).output();
+        }
+
+        let output = output.await?;
+
+        // TODO: handle termination conditions without defaulting to 0
+        let res = FidError {
+            fid: fid.fid.clone(),
+            data: fid.data.clone(),
+            errno: output.status.code().unwrap_or(0) as i16,
+        };
+        result.push(res);
+
+        tracing::debug!(
+            "exited with {} {} {}",
+            output.status,
+            std::str::from_utf8(&output.stdout)?,
+            std::str::from_utf8(&output.stderr)?
+        );
+    }
+
+    Ok(result)
+}
+
+/// Process FIDs
+pub async fn process_fids(
+    (fsname_or_mntpath, task_args, fid_list): (String, HashMap<String, String>, Vec<FidItem>),
+) -> Result<Vec<FidError>, ImlAgentError> {
+    let llapi = search_rootpath(fsname_or_mntpath).await?;
+
+    let dest_path = task_args
+        .get("remote".into())
+        .ok_or(RequiredError("Task missing 'remote' argument".to_string()))?;
+
+    let action_name = task_args
+        .get("action".into())
+        .ok_or(RequiredError("Task missing 'action' argument".to_string()))?;
+
+    // fuck it, panic.  I can'd figure out how to return an error from a match
+    // statement, but I can above.
+    match action_name.as_str() {
+        "push" => archive_fids(llapi.clone(), dest_path, fid_list).await,
+        "pull" => restore_fids(llapi.clone(), dest_path, fid_list).await,
+        _ => panic!(),
+    }
+}
```

Comment:
```suggestion
#[derive(Debug, serde::Deserialize, Clone, Copy)]
#[serde(rename_all = "lowercase")]
pub enum ActionType {
    Push,
    Pull,
}

#[derive(Debug, serde::Deserialize)]
pub struct TaskArgs {
    pub remote: String,
    pub action: ActionType,
}

/// Process FIDs
pub async fn process_fids(
    (fsname_or_mntpath, task_args, fid_list): (String, TaskArgs, Vec<FidItem>),
) -> Result<Vec<FidError>, ImlAgentError> {
    let llapi = search_rootpath(fsname_or_mntpath).await?;

    match task_args.action {
        ActionType::Push => archive_fids(llapi.clone(), &task_args.remote, fid_list).await,
        ActionType::Pull => restore_fids(llapi.clone(), &task_args.remote, fid_list).await,
    }
}
```

This does the exact same thing as before but the deserialization is handled by the compiler instead of you having to write lookup code.

Reply:
Done

Reply:
Done